OneCommunityGlobal/HGNRest
# Jerry - streak fix auto dropoff

# Description
This PR aims to solve this bug:
<img width="743" alt="Screenshot 2023-12-22 at 1 07 03 PM" src="https://github.com/OneCommunityGlobal/HGNRest/assets/55968519/0d08fb10-547d-4486-82c2-2d912b7afe1d">


## Related PRS (if any):
None.

## Main changes explained:
- Ensured that `changeBadgeCount()` does not get passed undefined value by changing `badgeCollection[i].badge.count - 1` to `badgeCollection[i].count - 1`
- When checking ‘X Hours for X Week Streak’ badges that are greater than 1 week long, ensured that the badges are checked in descending order by `week` and `totalHrs` values. This ensures that the auto dropoff logic works correctly. E.g., 30 Hours for 2 Weeks Streak badge is replaced by a 30 Hours for 3 Weeks Streak badge, when previously the 30 Hours for 2 Weeks Streak badge would not be removed and the user will keep that badge in addition to the newly awarded 30 Hours for 3 Weeks Streak badge.
- Added logic to correctly award the same type of streak badge to the user that they already possess. 
- Added logic to correctly avoid awarding streak badges to a user when “in between streak badges” and building upon an ongoing streak. E.g., a user that logs four consecutive weeks of 30 hours is awarded with the “30 Hours for 4 Weeks Streak” badge. The next streak badge to earn is “30 Hours for 6 Weeks Streak” badge. Therefore, when the user logs their fifth consecutive week of 30 hours, the logic detects that the user is building upon an ongoing streak and will not award the user with another “30 Hours for 4 Weeks Streak” badge. 

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. For testing purposes, install jest by running this command `npm i jest`. You may uninstall jest after you finish testing.
4. Inside package.json, change a line under “scripts”. Change “test”: “test” to “test”: “jest”
5. Run the test file by running this command: `npm test userHelper.test.js`
6. These unit tests take a relatively long time to run (around 10-15 seconds on my computer for bigger tests). If any test fails due to timing out, ajust the second argument provided to the `test()` calls. E.g., 10000 equals 10 seconds until timeout. Increase that number accordingly.
7. `ctrl` + `c` to force stop the jest execution.
8. Note: you may uninstall jest after testing this PR if you want.

After understanding the bug, you may try to come up with alternative test cases that may not have been considered yet. You can create your own test case by copying the structure of any existing test case in userHelper.test.js. Please keep in mind the X Hours For X Weeks badges that exist in the database when thinking of alternative test cases. A spreadsheet containing every badge can be found [here](https://docs.google.com/spreadsheets/d/1zq7BY7k8EB6_4sO8Lo5M8A2bOhwNBPQmUFmtT66cauI/edit?usp=sharing). By coming up with new test cases, we can ensure that the code changes are correct and robust.

### How to run test file:

https://github.com/OneCommunityGlobal/HGNRest/assets/55968519/0806df82-6583-403d-a4d5-a48eda5dc517

## Screenshots or videos of changes:

How to use Visual Studio Code debugger to debug badge component:

https://youtu.be/GYR6GSAVC_k
